miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultArrayRunService.java

```diff
@@ -0,0 +1,239 @@
+    ArrayRun managed = get(arrayRun.getId());
+    validateChange(arrayRun, managed);
+    applyChanges(arrayRun, managed);
+    setChangeDetails(managed);
+    return arrayRunStore.save(managed);
+  }
+
+  /**
+   * Checks submitted data for validity, throwing a ValidationException containing all of the errors if invalid
+   * 
+   * @param arrayRun submitted Array Run to validate
+   * @param beforeChange the already-persisted Array Run before changes
+   * @throws IOException
+   */
+  private void validateChange(ArrayRun arrayRun, ArrayRun beforeChange) throws IOException {
+    List<ValidationError> errors = new ArrayList<>();
+
+    if (arrayRun.getId() != ArrayRun.UNSAVED_ID && beforeChange == null) {
+      errors.add(new ValidationError("Array Run not found"));
+    }
+    if (isStringEmptyOrNull(arrayRun.getAlias())) {
+      errors.add(new ValidationError("alias", "Alias cannot be blank"));
+    }
+    if (arrayRun.getInstrument() == null) {
+      errors.add(new ValidationError("instrument", "An instrument must be selected"));
+    } else if (arrayRun.getInstrument().getPlatform().getInstrumentType() != InstrumentType.ARRAY_SCANNER) {
+      errors.add(new ValidationError("instrument", "Instrument must be an array scanner"));
+    }
+
+    if (beforeChange == null || !arrayRun.getAlias().equals(beforeChange.getAlias())) {
+      ArrayRun duplicateAlias = arrayRunStore.getByAlias(arrayRun.getAlias());
+      if (duplicateAlias != null) {
+        errors.add(new ValidationError("alias", "There is already an array run with this alias"));
+      }
+    }
+
+    if (arrayRun.getArray() != null && arrayRun.getArray().getAlias() == null) {
+      errors.add(new ValidationError("array", "Array not found"));
+    }
+
+    validateStartDate(arrayRun, beforeChange, errors);
+    validateStatusAndCompletion(arrayRun, beforeChange, errors);
+
+    if (!errors.isEmpty()) {
+      throw new ValidationException(errors);
+    }
+  }
+
+  private void validateStartDate(ArrayRun arrayRun, ArrayRun beforeChange, List<ValidationError> errors) throws IOException {
+    if (arrayRun.getStartDate() == null) {
+      errors.add(new ValidationError("startDate", "Start date cannot be blank"));
+    } else if (!authorizationManager.isAdminUser() && beforeChange != null && beforeChange.getStartDate() != null
+        && !arrayRun.getStartDate().equals(beforeChange.getStartDate())) {
+      errors.add(new ValidationError("startDate", "Only admin may change start date"));
+    }
+  }
+
+  private void validateStatusAndCompletion(ArrayRun arrayRun, ArrayRun beforeChange, List<ValidationError> errors) throws IOException {
+    if (arrayRun.getHealth() == null) {
+      errors.add(new ValidationError("status", "A status must be selected"));
+    } else if (arrayRun.getHealth().isDone()) {
+      if (arrayRun.getCompletionDate() == null) {
+        errors.add(new ValidationError("completionDate", "Completion date must be entered for a completed run"));
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "completionDate" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)